src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/Formatting/RemoteFormattingService.cs

```diff
@@ -0,0 +1,175 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor;
+using Microsoft.CodeAnalysis.Razor.Formatting;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.CodeAnalysis.Razor.Remote;
+using Microsoft.CodeAnalysis.Remote.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Response = Microsoft.CodeAnalysis.Razor.Remote.IRemoteFormattingService.TriggerKind;
+
+namespace Microsoft.CodeAnalysis.Remote.Razor;
+
+internal sealed class RemoteFormattingService(in ServiceArgs args) : RazorDocumentServiceBase(in args), IRemoteFormattingService
+{
+    internal sealed class Factory : FactoryBase<IRemoteFormattingService>
+    {
+        protected override IRemoteFormattingService CreateService(in ServiceArgs args)
+            => new RemoteFormattingService(in args);
+    }
+
+    private readonly IRazorFormattingService _formattingService = args.ExportProvider.GetExportedValue<IRazorFormattingService>();
+
+    public ValueTask<ImmutableArray<TextChange>> GetDocumentFormattingEditsAsync(
+        RazorPinnedSolutionInfoWrapper solutionInfo,
+        DocumentId documentId,
+        ImmutableArray<TextChange> htmlChanges,
+        RazorFormattingOptions options,
+        CancellationToken cancellationToken)
+        => RunServiceAsync(
+            solutionInfo,
+            documentId,
+            context => GetDocumentFormattingEditsAsync(context, htmlChanges, options, cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<ImmutableArray<TextChange>> GetDocumentFormattingEditsAsync(
+        RemoteDocumentContext context,
+        ImmutableArray<TextChange> htmlChanges,
+        RazorFormattingOptions options,
+        CancellationToken cancellationToken)
+    {
+        var sourceText = await context.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+        var htmlEdits = htmlChanges.Select(sourceText.GetTextEdit).ToArray();
+
+        var edits = await _formattingService.GetDocumentFormattingEditsAsync(context, htmlEdits, range: null, options, cancellationToken).ConfigureAwait(false);
+
+        if (edits is null)
+        {
+            return ImmutableArray<TextChange>.Empty;
```

Comment:
Consider a collection expression here, since you did that elsewhere in this file.

```suggestion
            return [];
```